File: examples/delphi/my-app/src/Controllers/UserController.pas
unit Controllers.UserController;

interface

uses
  System.SysUtils,
  Models.User,
  Services.UserService,
  Utils.Logger;

type
  TUserController = class
  private
    FService: TUserService;
    FLogger: TLogger;
  public
    constructor Create;
    destructor Destroy; override;
    
    function GetUserByUsername(const AUsername: string): TUser;
    function GetAllUsers: TArray<TUser>;
    function CreateUser(const AUsername, AEmail, APassword: string): Boolean;
    function DeactivateUser(AUserId: Integer): Boolean;
  end;

implementation

constructor TUserController.Create;
begin
  inherited;
  FService := TUserService.Create;
  FLogger := TLogger.Create;
end;

destructor TUserController.Destroy;
begin
  FService.Free;
  FLogger.Free;
  inherited;
end;

function TUserController.GetUserByUsername(const AUsername: string): TUser;
begin
  Result := FService.GetUserByUsername(AUsername);
end;

function TUserController.GetAllUsers: TArray<TUser>;
begin
  Result := FService.GetAllUsers;
end;

function TUserController.CreateUser(const AUsername, AEmail, APassword: string): Boolean;
begin
  try
    FService.CreateUser(AUsername, AEmail, APassword);
    Result := True;
  except
    on E: Exception do
    begin
      FLogger.LogError('Create user failed: ' + E.Message);
      Result := False;
    end;
  end;
end;

function TUserController.DeactivateUser(AUserId: Integer): Boolean;
begin
  try
    FService.DeactivateUser(AUserId);
    Result := True;
  except
    on E: Exception do
    begin
      FLogger.LogError('Deactivate user failed: ' + E.Message);
      Result := False;
    end;
  end;
end;

end.

File: examples/kotlin/my-app/src/main/kotlin/com/example/controllers/UserController.kt
package com.example.controllers

import com.example.models.User
import com.example.services.UserService

class UserController {
    private val userService = UserService()
    
    fun createUser(email: String, name: String, password: String = "password123"): User {
        return userService.createUser(email, name, password)
    }
    
    fun getUser(id: String): User? {
        return userService.getUserById(id)
    }
    
    fun updateUser(id: String, name: String): User? {
        return userService.updateUser(id, name)
    }
    
    fun deleteUser(id: String): Boolean {
        return userService.deleteUser(id)
    }
}
